webpack/assets/javascripts/react_app/components/common/forms/ForemanForm/ForemanForm.js

```diff
@@ -18,42 +25,71 @@ export const prepareErrors = errors =>
 export const isInitialValid = ({ validationSchema, initialValues }) =>
   !validationSchema ? true : validationSchema.isValidSync(initialValues);
 
-const ForemanForm = props => (
-  <Formik
-    onSubmit={(values, actions) =>
-      props.onSubmit(values, actions).catch(exception => {
-        actions.setSubmitting(false);
-        actions.setErrors(prepareErrors(exception.errors));
-      })
-    }
-    initialValues={props.initialValues}
-    validationSchema={props.validationSchema}
-    isInitialValid={isInitialValid}
-    enableReinitialize={props.enableReinitialize}
-  >
-    {formProps => {
-      const disabled = formProps.isSubmitting || !formProps.isValid;
+const ForemanForm = ({
+  onSubmit,
+  onSuccess,
+  children,
+  initialValues,
+  validationSchema,
+  enableReinitialize,
+  onCancel,
+}) => {
+  const [key, setKey] = useState(undefined);
+  const error = useSelector(state => selectAPIError(state, key));
+  const status = useSelector(state => selectAPIStatus(state, key));
+  const [formActions, setFormActions] = useState();
 
-      const submissionError = formProps.errors._error;
-      return (
-        <Form
-          onSubmit={formProps.handleSubmit}
-          onCancel={props.onCancel}
-          disabled={disabled}
-          error={submissionError}
-          errorTitle={
-            submissionError && submissionError.severity === 'danger'
-              ? __('Error! ')
-              : __('Warning! ')
-          }
-          submitting={formProps.isSubmitting}
-        >
-          {cloneChildren(props.children, { formProps, disabled })}
-        </Form>
+  useEffect(() => {
+    debugger;
+    if (key && formActions && error) {
+      formActions.setSubmitting(false);
+      formActions.setErrors(
+        prepareErrors(error?.response?.data?.error?.errors)
       );
-    }}
-  </Formik>
-);
+      setKey(undefined);
+    }
+  }, [key, formActions, error]);
+
+  useEffect(() => {
+    if (status === STATUS.RESOLVED) {
+      onSuccess && onSuccess();
+    }
+  }, [key, status]);
+  return (
+    <Formik
+      onSubmit={async (values, actions) => {
+        setKey(await onSubmit(values, actions));
+        setFormActions(actions);
+      }}
+      initialValues={initialValues}
+      validationSchema={validationSchema}
+      isInitialValid={isInitialValid}
+      enableReinitialize={enableReinitialize}
+    >
+      {formProps => {
+        const disabled = formProps.isSubmitting || !formProps.isValid;
+
+        const submissionError = formProps.errors._error;
```

Comment:
These need to be properly pulled from response as well:
![base_error](https://user-images.githubusercontent.com/2664090/107497386-cb96ff80-6b92-11eb-8698-efa322ccb22c.png)

Reply:
I guess now error handling should move from the forms action into the `ForemanForm` component as it will receive the data and decide what to show based on status, etc...

Reply:
@xprazak2, I didn't change here anything, only the location has changed.
`errors._error` isn't pulled from the response?

Reply:
It is pulled from the response correctly and the `useEffect` hook calls `setErrors`, but the errors are still absent. I am guessing formActions do not count on being stored in state of other component.

Looking at related [discussion](https://github.com/formium/formik/issues/33#issuecomment-314160295), we could probably pass `setErrors` as a middleware callback to be executed on error:

```js

onSubmit=(values, actions) => submitForm({
url,
 // other values
  actions
})

// redux/actions/common/forms.js

const submitForm({
 // ...
actions
}) => {
 // ...

dispatch(APIActions[method])({
// ...
handleError: response => actions.setErrors(...)
})
}
```

What do you think about this approach?